src/control/lib/control/rpc.go

```diff
@@ -488,6 +490,29 @@ func invokeUnaryRPC(parentCtx context.Context, log debugLogger, c UnaryInvoker,
 				return nil, err
 			}
 
+			// b/352550886: Add a forced-compatibility shim for Jumpbox code to be able to
+			// perform admin RPCs regardless of server version.
+			//
+			// NB: This workaround does not guarantee compatibility between admin:server
+			// versions! In most cases, calling admin RPCs from different versions should
+			// work fine, but there may be additional work needed to handle version differences.
+			if fault.IsFaultCode(err, code.ServerIncompatibleComponents) {
+				if f, ok := err.(*fault.Fault); ok {
+					verRe := regexp.MustCompile(fmt.Sprintf(`%s:(\d+\.\d+\.\d+)`, build.ComponentServer))
+					if matches := verRe.FindStringSubmatch(f.Description); len(matches) > 1 {
+						srvVer := matches[1]
+						log.Debugf("setting %s version to %s and retrying", c.GetComponent(), srvVer)
+						var ctxErr error
+						reqCtx, ctxErr = build.ToContext(reqCtx, c.GetComponent(), srvVer)
+						if ctxErr != nil {
+							log.Debugf("failed to set override version: %v", ctxErr)
+							return nil, errors.Wrapf(err, "unable to override %s version to match server's", c.GetComponent())
+						}
+						break
+					}
+				}
+			}
```

Comment:
this seems like quite a lot of code compared to other cases - do consider extracting it into a separate method with a descriptive name? e.g. detectClinetServerIncompatibility() and move all the comments into it. This would simplify this method significantly.

Reply:
Thanks for the review, Pavel! In general, I agree with you. I think there's a balance between reducing cyclomatic complexity and increasing readability, though. The logic here is specific to rpc error handling, and if I move it into a helper, then the reader has to jump to that helper to understand this error case.

That having been said, this section of the rpc processing code has grown organically over time, so it's probably due for a refactor at some point.